workflow/workflow_test.go

```diff
@@ -1213,6 +1213,124 @@ func TestInputCancelledStepWorkflow(t *testing.T) {
 	assert.Contains(t, err.Error(), "cannot construct any output")
 }
 
+var inputDisabledStepWorkflow = `
+version: v0.2.0
+input:
+  root: WorkflowInput
+  objects:
+    WorkflowInput:
+      id: WorkflowInput
+      properties:
+        step_enabled:
+          type:
+            type_id: bool
+steps:
+  simple_wait:
+    plugin:
+      src: "n/a"
+      deployment_type: "builtin"
+    step: wait
+    deploy:
+      deployer_name: "test-impl"
+      # Disabling steps waits for the value to be resolved before executing the step.
+      # So if it's working correctly, this should not cause any race conditions.
+      deploy_time: 0 # ms
+    input:
+      wait_time_ms: 20
+    enabled: !expr $.input.step_enabled
+outputs:
+  success:
+    simple_wait_output: !expr $.steps.simple_wait.outputs.success
+`
+
+func TestInputDisabledStepWorkflow(t *testing.T) {
+	// Run a workflow with one step that has its enablement state
+	// set by the input. The output only passes on success, so
+	// this test case also tests that the failure case doesn't
+	// lead to a deadlock.
+	preparedWorkflow := assert.NoErrorR[workflow.ExecutableWorkflow](t)(
+		getTestImplPreparedWorkflow(t, inputDisabledStepWorkflow),
+	)
+	// The workflow should pass with it enabled
+	outputID, _, err := preparedWorkflow.Execute(context.Background(), map[string]any{
+		"step_enabled": true,
+	})
+	assert.NoError(t, err)
+	assert.Equals(t, outputID, "success")
+	// The workflow should fail with it disabled
+	_, _, err = preparedWorkflow.Execute(context.Background(), map[string]any{
+		"step_enabled": false,
+	})
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "cannot construct any output")
+}
+
+var dynamicDisabledStepWorkflow = `
+version: v0.2.0
+input:
+  root: WorkflowInput
+  objects:
+    WorkflowInput:
+      id: WorkflowInput
+      properties:
+        sleep_time:
+          type:
+            type_id: integer
+steps:
+  initial_wait:
+    plugin:
+      src: "n/a"
+      deployment_type: "builtin"
+    step: wait
+    input:
+      wait_time_ms: !expr $.input.sleep_time # ms
+  toggled_wait:
+    plugin:
+      src: "n/a"
+      deployment_type: "builtin"
+    step: wait
+    input:
+      wait_time_ms: 0
+    enabled: !expr $.steps.initial_wait.outputs.success.message == "Plugin slept for 20 ms."
+outputs:
+  success:
+    initial_wait_output: !expr $.steps.initial_wait.outputs.success
+    toggled_wait_output: !expr $.steps.toggled_wait.outputs.success
+  disabled:
+    initial_wait_output: !expr $.steps.initial_wait.outputs.success
+    toggled_wait_output: !expr $.steps.toggled_wait.disabled.output
+`
+
+func TestDelayedDisabledStepWorkflow(t *testing.T) {
+	// Run a workflow where the step is disabled by a value that isn't there at
+	// the start of the workflow; in this case from another step's output.
```

Comment:
Replace "there" with "available"?

Also, the text following the semicolon isn't actually an independent clause (or a clause, at all).  I suggest replacing the semicolon with a comma and "because the value comes from another step's output."

(You can take the comma from line 1307, where it's superfluous.  😉)